File: 조준장/level3/단어 변환.py
from collections import deque

# 한 번에 한 개의 알파벳만 바꿨을 때 가능한 단어
def change_word(x, y):
    cnt = 0        
    for i in range(len(x)):
        if x[i] != y[i]:
            cnt += 1
            
    if cnt == 1:
        return True
    return False

def solution(begin, target, words):
    
    if target not in words:
        return 0
    
    answer = 1e9
    n = len(words)
    visited = [False] * n
    queue = deque([[begin, 0]])

    while queue:
        word, cnt = queue.popleft()
    
        if word == target:
            answer = min(answer, cnt)
            
        for i in range(n):
            if change_word(word, words[i]) and not visited[i]:
                visited[i] = True
                queue.append([words[i], cnt+1])
    
    return answer

File: 조준장/level3/단어 변환.kt
import kotlin.math.min

class Solution {
    fun solution(begin: String, target: String, words: Array<String>): Int {
        
        if (!words.contains(target)) return 0
        
        var answer = 1e9.toInt()
        var queue = ArrayList<Pair<String, Int>>()
        val visited = MutableList<Boolean>(words.size){false}
        
        queue.add(Pair(begin, 0))
        
        while (!queue.isEmpty()){
            val (newBegin, cnt) = queue.removeAt(0)
            
            if (newBegin == target) answer = min(answer, cnt)
            
            words.indices.forEach{
                if (wordCheck(newBegin, words[it]) && !visited[it]){
                    visited[it] = true
                    queue.add(Pair(words[it], cnt+1))
                }
            }
        }
        
        return answer
    }
    
    fun wordCheck(begin: String, target: String) : Boolean {
        var cnt : Int = 0
        begin.indices.forEach{
            if (begin[it] != target[it]) cnt ++ 
        }
        return if (cnt == 1) true else false 
    }
}
